### PR TITLE
Added a flag to add more intialisms

### LINKED ISSUE #1344: How to generate a correct struct file name from "#/definitions/IPAM"
## Problem statement

Hi, All, 

I ran into one situation that in my swagger.yml, I have defined an object `IPAM` in https://github.com/alibaba/pouch/blob/master/apis/swagger.yml#L1713-L1726:

```
  IPAM:
    type: "object"
    description: "represents IP Address Management"
    properties:
      Driver:
        type: "string"
      Options:
        type: "object"
        additionalProperties:
          type: "string"
      Config:
        type: "array"
        items:
          $ref: '#/definitions/IPAMConfig'
```

When generating a file, a file named `ip_a_m.go` is generated, see https://github.com/alibaba/pouch/blob/master/apis/types/ip_a_m.go.

I do not think the file name `ip_a_m.go` is what I want. 

So, could I customize the file name in some places, like setting something to generate a file named `IPAM.go`?

/cc @rudyfly @CodeJuan 

## Swagger specification
See https://github.com/alibaba/pouch/blob/master/apis/swagger.yml

## Steps to reproduce
NONE

## Environment
swagger version: `version: 0.12.0`
go version: `go version go1.9.1 darwin/amd64`
OS:   `Darwin AllenSundeMacBook-Pro.local 16.7.0 Darwin Kernel Version 16.7.0: Mon Nov 13 21:56:25 PST 2017; root:xnu-3789.72.11~1/RELEASE_X86_64 x86_64`

### PR DESCRIPTION
**Pull request by diego-fu-hs**
Added a flag to take in additional initialisms

Fixes #1344

### CODE REVIEW

**Comment by diego-fu-hs** [comment]
I am not sure why it is failing to build?

File: docs/generate/client.md [hunk 1]
```diff
@@ -35,6 +35,7 @@ Help Options:
           --dump-data         when present dumps the json for the template generator instead of generating files
           --skip-validation   skips validation of spec prior to generation
       -r, --copyright-file=   the file containing a copyright header for the generated source
+	  --additional-initiliasm=  additional consecutive capitals that should be considered as initialism, repeat for multiple
```
Comment (fredbi) [inline docs/generate/client.md]: typo in option here

File: docs/generate/server.md [hunk 2]
```diff
@@ -47,6 +47,7 @@ Help Options:
           --compatibility-mode=[modern|intermediate] the compatibility mode for the tls server (default: modern)
           --skip-validation                          skips validation of spec prior to generation
       -r, --copyright-file=                          the file containing a copyright header for the generated source
+	  --additional-initiliasm=                      additional consecutive capitals that should be considered as initialism, repeat for multiple
```
Comment (fredbi) [inline docs/generate/server.md]: typo in option here

**Comment by fredbi** [comment]
Couldn't see why exactly, but can point as to see where.
It's in `generator/build_test.go`
This simulates a go get and build.

**Comment by fredbi** [comment]
Perhaps upgrading the test to use `go get gopkg.in/go-swagger/go-swagger.{current-branch}` and retrieve the current branch from the build env?

**Comment by diego-fu-hs** [comment]
@fredbi found what the problem was.

Do you mind taking another look? Thanks!

File: cmd/swagger/commands/generate/support.go [hunk 3]
```diff
@@ -45,39 +43,26 @@ func (s *Support) Execute(args []string) error {
...
-	}
-	rp, err = filepath.Rel(basepath, targetAbs)
-	if err != nil {
-		return err
-	}
+func (s *Support) log(rp string) {
 
 	log.Printf(`Generation completed!
 
-For this generation to compile you need to have some packages in your vendor or GOPATH:
```
Comment (casualjim) [inline cmd/swagger/commands/generate/support.go]: please retain the vendor folder here, it's not clear to new comers that they are the same

File: cmd/swagger/commands/generate/support.go [hunk 4]
```diff
@@ -45,39 +43,26 @@ func (s *Support) Execute(args []string) error {
...
+For this generation to compile you need to have some packages in your GOPATH:
 
-  * github.com/go-openapi/runtime
-  * github.com/asaskevich/govalidator
-  * github.com/tylerb/graceful
-  * github.com/jessevdk/go-flags
-  * golang.org/x/net/context/ctxhttp
+	* github.com/go-openapi/runtime
```
Comment (casualjim) [inline cmd/swagger/commands/generate/support.go]: please retain the list of packages, depending on which vendorint tool you use you need the exhaustive list

File: generator/shared.go [hunk 5]
```diff
@@ -247,7 +247,7 @@ func DefaultSectionOpts(gen *GenOpts, client bool) {
 	}
 
 	if len(sec.Operations) == 0 {
-		if client {
+		if gen.EnsureDefaultOpts {
```
Comment (casualjim) [inline generator/shared.go]: this variable name is very misleading because this is about picking a set of templates for either client or server gen.
I don't know why EnsureDefaultOpts should be the switch when it's already doing the default opts.

File: cmd/swagger/commands/generate/support.go [hunk 6]
```diff
@@ -45,39 +43,26 @@ func (s *Support) Execute(args []string) error {
...
 		DefaultScheme: s.DefaultScheme,
 		TemplateDir:   string(s.TemplateDir),
-	}
+	}, nil
+}
 
-	if err := generator.GenerateSupport(s.Name, nil, nil, &opts); err != nil {
-		return err
-	}
+func (s *Support) generate(opts *generator.GenOpts) error {
```
Comment (casualjim) [inline cmd/swagger/commands/generate/support.go]: I'd like to either see this method renamed or the package level generate method renamed. I had to do quite a bit of digging to work out that the config file is still being read

File: generator/shared.go [hunk 7]
```diff
@@ -406,6 +406,7 @@ type GenOpts struct {
 	ValidateSpec       bool
 	FlattenSpec        bool
 	FlattenDefinitions bool
+	EnsureDefaultOpts  bool
```
Comment (casualjim) [inline generator/shared.go]: please rename to something like IsClient or equivalent that expresses meaning

File: cmd/swagger/commands/generate/support.go (hunk 4 above)
Comment (diego-fu-hs) [reply cmd/swagger/commands/generate/support.go]: This was a copy-paste error :D

**Comment by fredbi** [comment]
I can see this time this is a spec validation test that failed.
Wonder if it is a regression introduced my latest PR with spec valition.
I'll test with this fixture (docker spec).
Strange, I specifically made a test to run the spec validate through all these fixtures in the go-openapi/validate package...

**Comment by fredbi** [comment]
I had a look at the fixtures/canary/docker/swagger.json specs which is blocking the UT:
it is considered invalid as for now for swagger.
It was already considered invalid when PR merged the new validate version.
I don't understand why this is blocking the UT now since this has been a known fact for some time.

Spec validation errors are acceptable for now (we ought to improve, though, so docker may use go-swagger :) ):
- support for multiple types, e.g: type: [ 'object', 'array' ] : this is a bug in validate but I don't think it is new
- duplicate operationId : this is a known limitation in go-swagger (operations must be globally unique while the spec states that (path,operationId) must be unique

**Comment by diego-fu-hs** [comment]
@fredbi anything I can do to fix those? I am pretty lost right now

**Comment by fredbi** [comment]
I think we need an explanation on how suddenly the UT on validation requires this docker spec to be valid.

It is, in a sense, but we can't support it yet. I must admist that this side effect of your otherwise unrelated change is hard to follow. I'll try to have a look over the we.

First PR I made I had the kind of bizarre stuff from CI. It is a complex application, indeed.

**Comment by casualjim** [comment]
I should look at the changes for the validation again. I suspect that there were things that use to be warnings that are now errors.

**Comment by fredbi** [comment]
@diego-fu-hs  : to pass the CI/CircleCI
- in ./hack/run-canary.sh we must disable the docker spec:
+ ` && [ $dir != 'docker' ] `

We have actually two issues which are related to my own previous PR and not yours. So I'll take care of fixing them.

**Comment by fredbi** [comment]
For the AppVeyor CI, I didn't find out. The fixture "bug/825" tested is expected to fail ok.
Why we have a timeout I don't know. I could see all tests where very long in this CI session. So maybe it was just unrelated. Let's run the CI again after a commit bump

File: hack/run-canary.sh [hunk 8]
```diff
@@ -17,10 +17,10 @@ do
   if [ $dir != "bitbucket.org" ]; then
     pushd "${FIXTURES}/$dir"
     rm -rf client models restapi cmd
-    swagger generate client --skip-validation
+    swagger generate client
```
Comment (fredbi) [inline hack/run-canary.sh]: This change has change the CI's result on CircleCI docker build task:
skipping th validation alllowed to pass docker - now validation is enforced so we can't generate

File: hack/run-canary.sh [hunk 9]
```diff
@@ -17,10 +17,10 @@ do
...
     go test ./...
     if [ $dir != 'kubernetes' ] && [ $dir != 'ms-cog-sci' ] ; then
-      swagger generate server --skip-validation
+      swagger generate server
```
Comment (fredbi) [inline hack/run-canary.sh]: Same here

This change has change the CI's result on CircleCI docker build task:
skipping th validation alllowed to pass docker - now validation is enforced so we can't generate

File: generator/server_test.go [hunk 10]
```diff
@@ -149,7 +149,6 @@ func TestServer_MultipartForm(t *testing.T) {
 func TestServer_InvalidSpec(t *testing.T) {
 	opts := testGenOpts()
 	opts.Spec = "../fixtures/bugs/825/swagger.yml"
-	opts.ValidateSpec = true
```
Comment (fredbi) [inline generator/server_test.go]: This has caused the CI Appveyor to hang I believe.
We should not let the generation run on this spec : the Validate spec is I believe important

**Comment by fredbi** [comment]
OK I understand better now. I suppose you have discussed with @casualjim about removing the --skip-validation flag.
However it has some impact on the CI, since some tests where specifically running in that grey zone of invalid specs for which a generation could be run...

+ canary/docker (in ./hack/run-canary.sh) was one of them
+ I suppose that the spec in ./fixture/bug/825 (which does not validate) is like this (see run failure on AppVeyor)

In https://github.com/go-openapi/validate/blob/master/swagger_test.go there is a full list of all fixtures from the go-swagger repo that do not validate [used to test the validate package itself].

**Comment by casualjim** [comment]
o i didn't mean skip-validation but skip-validator flag.

we want to validate but the skip-validator we don't want to support

**Comment by fredbi** [comment]
Ok, if I recap:
- `--skip-validation` has an impact on the docker build (circle ci) since we attempt to generate specs that are not . This one should not be removed.
- `--skip-validation` is not used and neither useful (?). Removing it from implem and doc should "fix" #1354.

Now about the `go-openapi/validate` package, this P.R inspired me some todos there (:imp:):
- [ ] Support the construct: `- type: [ xxx, array ]` when checking every `items` refers to an array
- [ ] Simulate a go-swagger merge in Travis and run the docker-build script. This would be an non-blocking Travis CI job, just to check that a validate branch may be merge into go-swagger without harming the CI.

**Comment by fredbi** [comment]
Fixes #1354

**Comment by casualjim** [comment]
swagger doesn't allow for type arrays so they dont' need to be supported

**Comment by casualjim** [comment]
we should remove the makefile and the scripts that uses. they are not useful and confusing

**Comment by fredbi** [comment]
So what do you suggest we use for C.I? As for now, Circle CI actually checks it can build a docker image.

If you remove the canary hurdle from go-swagger, I think I'll add it on the validate side.
For validation, I think it's always a good idea to benchmark against big such big specs

**Comment by diego-fu-hs** [comment]
oh wow thanks guys.

I totally mixed up `skip-validation` and `skip-validator` flags......

**Comment by diego-fu-hs** [comment]
@casualjim do you mind taking another look since i took out the wrong flag? Thanks!

**Comment by fredbi** [comment]
@diego-fu-hs as far as I can see it looks okay now. What do you think @casualjim ?

**Comment by casualjim** [comment]
it needs a rebase on master because there was another PR merged with changes to templates, so bindata is in conflict currently

**Comment by diego-fu-hs** [comment]
Done!
